pkg/util/wait/wait.go

```diff
@@ -237,3 +237,53 @@ func poller(interval, timeout time.Duration) WaitFunc {
 		return ch
 	})
 }
+
+type Interval struct {
+	Count    int
+	Interval time.Duration
+}
+
+func NewChannelDelayedAction(steps ...Interval) *ChannelDelayedAction {
+	return &ChannelDelayedAction{steps: steps}
+}
+
+type ChannelDelayedAction struct {
+	active <-chan time.Time
+	t      *time.Timer
+
+	steps    []Interval
+	count    int
+	interval time.Duration
+}
+
+func (q *ChannelDelayedAction) After() <-chan time.Time {
+	return q.active
+}
+
+func (q *ChannelDelayedAction) Done() {
+	q.active = nil
+}
+
+func (q *ChannelDelayedAction) Start() {
+	if q.active != nil {
+		q.count++
+		return
+	}
+	q.updateInterval()
+	q.count = 0
+	if q.t == nil {
+		q.t = time.NewTimer(q.interval)
+	} else if q.t.Reset(q.interval) {
+		<-q.t.C
```

Comment:
When you reset a timer, if it returns true you have to wait for a tick on
the channel (otherwise you will leak goroutines).

On Tue, Apr 26, 2016 at 7:26 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/util/wait/wait.go
> <https://github.com/kubernetes/kubernetes/pull/24768#discussion_r61069035>
> :
>
> > +
> > +func (q *ChannelDelayedAction) Done() {
> > +	q.active = nil
> > +}
> > +
> > +func (q *ChannelDelayedAction) Start() {
> > +	if q.active != nil {
> > +		q.count++
> > +		return
> > +	}
> > +	q.updateInterval()
> > +	q.count = 0
> > +	if q.t == nil {
> > +		q.t = time.NewTimer(q.interval)
> > +	} else if q.t.Reset(q.interval) {
> > +		<-q.t.C
>
> hmm - why we are waiting here for the whole interval?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24768/files/91a0c160abd3ac01d0dbe9a89b3b5131c0cad419#r61069035>
>